### PR TITLE
feat: contact verification request

### PR DESCRIPTION
**Pull request by richard-ramos**
- [x] Send contact verification request `wakuext_sendContactVerificationRequest(contactID, challenge)`
- [x] Accept contact verification request `wakuext_acceptContactVerificationRequest(contactID, response)`
- [x] Decline contact verification request `wakuext_declineContactVerificationRequest(contactID)`
- [x] Cancel contact verification request `wakuext_cancelVerificationRequest(contactID)`
- [x] Mark as trusted `wakuext_markAsTrusted(contactID)`
- [x] Mark as untrustworthy `wakuext_markAsUntrustworthy(contactID)`
- [x] Remove trust status `wakuext_removeTrustStatus(contactID)`
- [x] Get trust status `wakuext_getTrustStatus(contactID)`
- [ ] Create activity center notifications
- [x] Test units

### CODE REVIEW

**Comment by ghost** [comment]
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

File: protocol/contact.go [hunk 1]
```diff
@@ -55,6 +56,14 @@ func (c *Contact) CanonicalImage(profilePicturesVisibility accounts.ProfilePictu
 	return c.Identicon
 }
 
+type VerificationStatus int
+
+const (
+	VerificationStatusUNVERIFIED VerificationStatus = 0
+	VerificationStatusVERIFYING  VerificationStatus = 1
+	VerificationStatusVERIFIED   VerificationStatus = 2
+)
```
Comment (Samyoul) [inline protocol/contact.go]: Would you consider using `iota`, it seems like a nice fit for your enums

```suggestion
const (
	VerificationStatusUNVERIFIED VerificationStatus = iota
	VerificationStatusVERIFYING  VerificationStatus
	VerificationStatusVERIFIED   VerificationStatus
)
```

File: protocol/verification/persistence.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+	switch err {
+	case sql.ErrNoRows:
+		return nil, nil
+	case nil:
+		vr.RequestedAt = time.Unix(requestedAt, 0)
+		vr.RepliedAt = time.Unix(repliedAt, 0)
+		return &vr, nil
+	default:
+		return nil, err
+	}
```
Comment (Samyoul) [inline protocol/verification/persistence.go]: I like this

File: protocol/verification/persistence.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+		return err
+	}
+
+	numRows, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	if numRows == 0 {
+		return errors.New("verification request not found")
```
Comment (Samyoul) [inline protocol/verification/persistence.go]: Would you consider something like? If you've got multiple instances of the same error being returned, consuming code can check for a specific err via direct comparison `if err == ErrVerificationRequestNotFound { do something }`, also it saves you time if you need to call the same error in other places.

Obviously the `var()` block would go at the top in the package context.

```suggestion
var (
	ErrVerificationRequestNotFound = errors.New("verification request not found")
)

		return ErrVerificationRequestNotFound 
```

**Review by Samyoul — APPROVED** [review]
Great PR, a few points but just convention stuff.

File: protocol/messenger_contact_verification.go [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+	}
+
+	err = m.SyncVerificationRequest(context.Background(), persistedVR)
+	if err != nil {
+		return err
+	}
+
+	state.AllVerificationRequests = append(state.AllVerificationRequests, persistedVR)
+
+	contact.VerificationStatus = VerificationStatusVERIFIED
```
Comment (jrainville) [inline protocol/messenger_contact_verification.go:719]: I have a small issue where at the end of the process, the request is marked as trusted, but the verificationStatus stays at 0 when I look at it on the client side, maybe you'll be able to help me @richard-ramos

**Comment by churik** [comment]
@richard-ramos can you please resolve conflicts first and rebase to current develop? 
Thanks!

**Comment by jrainville** [comment]
This can be reviewed and merged when done

File: protocol/messenger.go [hunk 5]
```diff
@@ -5723,3 +5897,79 @@ func (m *Messenger) handleSyncClearHistory(state *ReceivedMessageState, message
...
+
+		contact.TrustStatus = verification.TrustStatus(message.Status)
+		m.allContacts.Store(contact.ID, contact)
+		state.ModifiedContacts.Store(contact.ID, true)
+	}
+
+	return nil
+}
+
+func ToVerificationRequest(message protobuf.SyncVerificationRequest) *verification.Request {
```
Comment (0x-r4bbit) [inline protocol/messenger.go:5924]: Would be cool if this went into `protocol/requests/...` instead for consistency's sake, but doesn't have to

File: protocol/messenger.go [hunk 6]
```diff
@@ -5723,3 +5897,79 @@ func (m *Messenger) handleSyncClearHistory(state *ReceivedMessageState, message
...
+
+		contact.VerificationStatus = VerificationStatus(message.VerificationStatus)
+		if err := m.persistence.SaveContact(contact, nil); err != nil {
+			return err
+		}
+
+		m.allContacts.Store(contact.ID, contact)
+		state.ModifiedContacts.Store(contact.ID, true)
+
+		// TODO: create activity center notif
```
Comment (0x-r4bbit) [inline protocol/messenger.go:5967]: Are these yet to be done as part of this PR or is has this been put on TODO on purpose for now?

File: protocol/messenger_contact_verification.go (hunk 4 above)
Comment (0x-r4bbit) [reply protocol/messenger_contact_verification.go:719]: Is this still relevant?

File: protocol/migrations/sqlite/1655309635_contact_verification.up.sql [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+CREATE TABLE IF NOT EXISTS verification_requests (
+  from_user TEXT,
+  to_user TEXT,
+  challenge TEXT NOT NULL,
+  requested_at INT NOT NULL DEFAULT 0,
+  response TEXT,
+  replied_at INT NOT NULL DEFAULT 0,
+  verification_status INT NOT NULL DEFAULT 0,
+  CONSTRAINT fromto_unique UNIQUE (from_user, to_user) ON CONFLICT REPLACE
+);
```
Comment (0x-r4bbit) [inline protocol/migrations/sqlite/1655309635_contact_verification.up.sql]: Need to ensure we update the timestamp of this file right before merging (and then `make generate` again)

**Review by 0x-r4bbit — COMMENTED** [review]
Massive work!!
Left some tiny comments, but really well done (also thanks for adding all those tests!)

File: protocol/messenger.go (hunk 6 above)
Comment (jrainville) [reply protocol/messenger.go:5967]: Activity center stuff will be done in another PR

File: protocol/messenger_contact_verification.go (hunk 4 above)
Comment (jrainville) [reply protocol/messenger_contact_verification.go:719]: Ah my bad, Richard helped me fix that

File: protocol/migrations/sqlite/1655309635_contact_verification.up.sql (hunk 7 above)
Comment (jrainville) [reply protocol/migrations/sqlite/1655309635_contact_verification.up.sql]: Done

**Pull request merged**